email/sender.go

```diff
@@ -0,0 +1,217 @@
+package email
+
+import (
+	"context"
+	"crypto/tls"
+	"net/url"
+	"regexp"
+	"strconv"
+	"time"
+
+	"github.com/NebulousLabs/skynet-accounts/database"
+	"gitlab.com/SkynetLabs/skyd/skymodules"
+	"gopkg.in/mail.v2"
+
+	"github.com/sirupsen/logrus"
+	"gitlab.com/NebulousLabs/errors"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+const (
+	// emailBatchSize defines the largest batch of emails we will try to send.
+	emailBatchSize = 10
+
+	// sleepBetweenScans defines how long the sender should sleep between its
+	// sweeps of the DB.
+	sleepBetweenScans = 10 * time.Second
+)
+
+var (
+	// From is the address we send emails from. It defaults to the user
+	// from DefaultConnectionURI but can be overridden by the ACCOUNTS_EMAIL_FROM
+	// environment variable.
+	From = "noreply@siasky.net"
+
+	// PortalAddress defines the URI where we can access our portal. Its value
+	// comes from the PORTAL_DOMAIN environment variable, preceded by the
+	// appropriate schema.
+	PortalAddress = "https://siasky.net"
+
+	// ServerLockID holds the name of the name of this particular server. Its
+	// value is controlled by the SERVER_DOMAIN entry in the .env file. If the
+	// SERVER_DOMAIN entry is empty or missing, the PORTAL_DOMAIN (preceded by
+	// schema) will be used instead.
+	ServerLockID string
+
+	// matchPattern extracts all relevant configuration values from an email
+	// connection URI
+	matchPattern = regexp.MustCompile("smtps://(?P<user>.*):(?P<password>.*)@(?P<server>.*):(?P<port>\\d*)(/\\??skip_ssl_verify=(?P<skip_ssl_verify>\\w*))?")
+)
+
+type (
+	// Sender is a daemon that periodically checks the DB for emails waiting to
+	// be sent and sends them.
+	Sender struct {
+		staticConfig emailConfig
+		staticCtx    context.Context
+		staticDB     *database.DB
+		staticDeps   skymodules.SkydDependencies
+		staticLogger *logrus.Logger
+	}
+
+	// emailConfig contains all configuration options we need in order to send
+	// an email
+	emailConfig struct {
+		User               string
+		Pass               string
+		Server             string
+		Port               int
+		InsecureSkipVerify bool
+	}
+)
+
+// NewSender returns a new Sender instance.
+func NewSender(ctx context.Context, db *database.DB, logger *logrus.Logger, deps skymodules.SkydDependencies, connURI string) (Sender, error) {
+	c, err := config(connURI)
+	if err != nil {
+		return Sender{}, errors.AddContext(err, "failed to parse email config")
+	}
+	return Sender{
+		staticConfig: c,
+		staticCtx:    ctx,
+		staticDB:     db,
+		staticDeps:   deps,
+		staticLogger: logger,
+	}, nil
+}
+
+// Start periodically scanning the database for email messages waiting to be
+// sent and sending them.
+func (s Sender) Start() {
+	go func() {
+		for {
+			select {
+			case <-s.staticCtx.Done():
+				return
+			default:
+			}
+			s.scanAndSend()
+			time.Sleep(sleepBetweenScans)
```

Comment:
This can just be added to the select statement

```
s.scanAndSend()
select {
case <-s.staticCtx.Done():
case <-time.After(sleepBetweenScans):
}
```

Reply:
Or we could use a `time.AfterFunc` here

Reply:
nit / follow up

Reply:
Sure thing. Changed in the upcoming PR
```go
		for {
			select {
			case <-s.staticCtx.Done():
				return
			case <-time.After(sleepBetweenScans):
				s.scanAndSend()
			}
		}
```